server/src/main/java/org/apache/druid/server/coordination/ServerManager.java

```diff
@@ -197,13 +197,16 @@ public ServerManager(
     }
 
     // segmentMapFn maps each base Segment into a joined Segment if necessary.
-    final Function<SegmentReference, SegmentReference> segmentMapFn = Joinables.createSegmentMapFn(
+    final Function<SegmentReference, SegmentReference> segmentMapFn = joinables.createSegmentMapFn(
         analysis.getPreJoinableClauses(),
-        joinableFactory,
         cpuTimeAccumulator,
         analysis.getBaseQuery().orElse(query)
     );
 
+    final Optional<byte[]> cacheKeyPrefix = analysis.isJoin()
```

Comment:
I am not expecting cache key computation to be a costly operation. do you think it should still be passed via a supplier?

Reply:
Hmm, what I want is having cache key computation in one place since it's easier to follow how the key is computed. I think my previous comment was not enough for that. How about passing `DataSourceAnalysis` and `Joinables` to `CachingQueryRunner` so that cache key computation can be solely done there?

Reply:
I remember now that what you are describing is how I started. But changed it later on so that the caching key is not computed for every segment participating in the query. Thats why I compute the cacheKeyPrefix in `getQueryRunnerForSegments` and the pass it on to each `SpecificSegmentQueryRunner`.

Reply:
I see. It sounds reasonable to me although the cache key computation should be cheap. I think we can introduce another class responsible for computing cache keys. `ServerManager` can create it and pass it to `CachingQueryRunner`. The new class will take `DatasourceAnalysis` and `Joinables` in its constructor, and have an interface providing a computed cache key for a given segment. I don't think this refactoring should be necessarily done in this PR, but it seems good to have. Are you interested in this refactoring as well?

Reply:
I think this piece of code is too small to be refactored into its own class as such. May be in future if the cache key computation becomes more complex.